---
title: "Phenotype Pipeline"
---
<div style="display: flex; justify-content: space-between; align-items: center;">
  <a href="./index.html">⬅️ Return to Homepage</a>
  <a href="./gwas.html">Go to Step 4 [Genome Wide Association Study] ➡️</a>
</div>

# Phenotype Pipeline

## Overview

This repository provides a containerized pipeline for QC and deriviation of phenotype data for subsequent GWAS analysis.
The dependencies for the pipeline are distributed as a .sif file that can be run using apptainer/singularity. 

Typically, we expect baseline measures of anthropometric traits for analysis. However, for studies with longitudinal measures without a single baseline measure, 
we suggest using the median value across the time-series. Age for the median value should either be captured or calculated.


## Usage

1. **Clone the pre-phasing/imputation repo from GitHub**
   ```
   git clone git@github.com:giant-consortium/phenotype_processing.git
   ```

2. **Navigate to `phenotype_processing/` directory**
   ```
   cd phenotype_processing/
   ```

3. **Edit `parameters.txt` to set filenames and options for your data.** This includes mapping column names to data labels that can be interpreted by the pipeline.

4. **Run the pipeline**. This will download the container if it does not already exist in the working directory. 
You should specify whether you would like to use Apptainer or Singularity to run the container.

    ```
    # Run container using Apptainer
    bash PHENOTYPE_PIPELINE.sh --apptainer
    
    # Run container using Singulairty
    bash PHENOTYPE_PIPELINE.sh --singularity
    ```

5. **Outputs** will be saved in the directory specified by `out_dir` in parameters.txt. Two sub-directories will be created:

  * `regenie_input/`
    * primary phenotype file
    * additional covariate file associated with genotyping / imputation process
    * subject lists to stratify individuals by ancestry and sex

  * `summaries/`
    * tabular summaries of phenotypes, split by each phenotype and ancestry
    * png file plotting histograms and box plots of distributions of pre-QC'd, QC'd, residuals, and rank normal values for GWAS

  * `logs/`
    * .txt files containing individuals removed due to QC of phenotype
    * .txt files containing individuals removed as missing data for phenotype or at least one covariate for the respective trait

## Details

### Overview of phenotype preparation for GWAS
<img src="Overview_Phenotype_Pipeline.png"/>

### 1. Files required
The phenotype pipeline expects 3 files for input to be specified in the `parameters.txt` file.

| Parameter | Description |
| --- | --- |
| `pheno_cov_file` | File containing all study variables for phenotype derivation (not genotype PCs)|
| `ancestry_file`  | File containing `IID` and `Ancestry_Prediction` to map individuals to one ancestry  |
| `pcs_file`       | First 20 within ancestry genetic principal components generated by FLASHPCA |

##### Phenotype + covariate file
This file should be a tab-delimited file with header that contains all variables for phenotype derivation for height, BMI and WHR, and all associated covariates. 
```
ID    Age   Sex  Height  Weight  Waist  Hip    CatVar1  CatVar2  CatVar3
P1    25    2    99      75.6    89     100    0        1        1
P2    85    1    170     89.8    91     98     1        1        2
P3    36    2    159     91.1    84     1020   0        2        3
...
```
##### Ancestry label file - from Individual and Genotype QC Pipeline
This file should be a tab-delimited file with the header as shown with ancestry group assignment provided for each individual [afr|amr|eas|eur|mid|sas]
```
IID     Ancestry_Prediction
P1      eur
P2      eur
P3      eur
...
```
##### Principal components file - from Individual and Genotype QC Pipeline
This file should be a tab-delimited file with 20 PCs (FLASHPCA output format)
```
IID   FID   PC1          PC2          PC3            ...    PC18         PC19         PC20          
P1    1     0.722770917  0.422132091  0.516251618    ...    0.05984563   0.74798664   0.310145635
P2    2     0.584095559  0.154674118  0.243537544    ...    0.059630973  0.708644846  0.863891645
P3    3     0.709639726  0.926419979  0.041459368    ...    0.613060395  0.315596588  0.719991385
...
```

### 2. General parameters related to study

| Parameter | Description | Optional |
| --- | :--- | --- |
| `out_dir`| Parent directory for REGENIE input and summaries | No |
| `study` | Name of Study | No |
| `id_col` | Column containing individual IDs | No |
| `sex_col` | Column containing sex labels | No  |
| `m_label` | Label for males | Yes |
| `f_label` | Label for females | Yes |
| `case_col`| Column containing case/control labels | Yes |
| `case_label` | Label for disease cases | Yes |
| `control_label`| Label for disease controls | Yes |
| `qt_covar_cols` | Comma delimited QT covariate list for phenotype adjustment |Yes |
| `cat_covar_cols`| Comma delimited categorical covariate list for phenotype adjustment| Yes |
| `geno_qt_covar_cols` | Comma delimited QT covariate list for adjustment at runtime | Yes |
| `geno_cat_covar_cols` | Comma delimited categorical covariate list for adjustment at runtime | Yes |



### 3. Derivation of phenotypes
If possible, please provide variables to derive BMI and WHR, as well as any pre-derived values for BMI and WHR:

| Phenotype | Variables to include if available |
| :---: | :---: |
| **HEIGHT** | height (cm) |
| **BMI** | height (cm), weight (kg), pre-derived BMI |
| **WHR** | waist circumference (cm), hip circumference (cm), pre-derived WHR |

If variables to derive BMI and/or WHR have been provided, the pipeline will prioritise derivation of the phenotypes and only use pre-derived values (if provided) to fill any missing data where possible. 
Note, derivation of BMI and WHR will include QC of variables used prior to QC of BMI and WHR. If variables are not available to derive BMI and/or WHR but variables exist with pre-derived values, these will be used. 
  
#### Quality control
Variables used to derive BMI and WHR will first undergo QC prior to derivation of BMI and WHR that will subsequently be QC'd again.

Details of quality control applied to variables can be found in the following table. 
Z-scores are calculated within ancestry-sex specific strata, or within ancestry-sex-case/control strata for BMI and WHR if case/control labels are provided. 

| Variable / Phenotype | Range Inclusion | Z-score Inclusion |
| :---: | :---: | :---: |
| **HEIGHT**    | 100 - 220 (cm)    | -5 < z < 5 |
| **BMI**       | 15 - 100          | -5 < z(log BMI) < 5 |
| **WHR**       | NA                | -5 < z < 5 |
| **WHRadjBMI** | NA                | -5 < z < 5 |
| **WEIGHT**    | NA                | -5 < z < 5 |
| **WAIST**     | 38.1 - 228.6 (cm) | -5 < z < 5 |
| **HIP**       | NA                | NA |
| **AGE**       | 18-110 (years)    | NA |


### 4. Parameters specific to phenotypes for GWAS
To account for the nature of EHR-based studies and other longitudinal studies,
two sets of variables may be provided for BMI and WHR if the variables related to the BMI GWAS and WHR GWAS differ from those related to WHRadjBMI GWAS. 
If the same variable is to be used, then you can specify the same column header for each
 

#### Parameters related to Height GWAS

| Parameter | Description | Optional |
| --- | --- | --- |
| `height_col` | Column containing height values (cm) | Yes |
| `age_height_col` | Column containing age at height values (years) | Yes |

#### Parameters related to BMI GWAS

We would prefer for BMI to be derived from weight and height but a parameter is provided to specify pre-derived BMI if this is not possible.

| Parameter | Description | Optional |
| --- | --- | --- |
| `weight_for_bmi_col` | Column containing weight values (kg) (used if height QC'd) | Yes |
| `bmi_col`    | Column containing BMI values (ignored if deriving BMI) | Yes |
| `age_bmi_col` | Column containing age related to BMI (years) | Yes |

### Parameters related to WHR GWAS

We would prefer for WHR to be derived from waist and hip but a parameter is provided to specify pre-derived WHR if this is not possible.

| Parameter | Description | Optional |
| --- | --- | --- |
| `waist_for_whr_col` | Column containing waist values (cm) | Yes |
| `hip_for_whr_col`   | Column containing hip values (cm) | Yes |
| `whr_col`    | Column containing WHR values (ignored if deriving WHR) | Yes |
| `age_whr_col` | Column containing age related to WHR (years) | Yes |

### Parameters related to WHRadjBMI GWAS

We would prefer for WHR and BMI to be derived from weight, height, waist and hip, but parameters are provided to specify pre-derived WHR and/or BMI if this is not possible.

| Parameter | Description | Optional |
| --- | --- | --- |
| `waist_for_whradjbmi_col` | Column containing waist values (cm) | Yes |
| `hip_for_whradjbmi_col`   | Column containing hip values (cm) | Yes |
| `whr_col`    | Column containing WHR values (ignored if deriving WHR) | Yes |
| `weight_for_whradjbmi_col` | Column containing weight values (kg) (used if height QC'd) | Yes |
| `bmi_for_whradjbmi_col`    | Column containing BMI values (ignored if deriving BMI) | Yes |
| `age_whradjbmi__col` | Column containing age related to WHR adjusted for BMI (years) | Yes |


## Outputs
All outputs will be directed to the parent output directory specififed by the `out_dir` parameter. A table of outputs within generated sub-directories is listed below:

| Subdirectory | Files | Descroption |
| --- | --- | --- |
| `regenie_input/` | `*.pheno`  | Tab-delimited file containing phenotypes to GWAS |
| `regenie_input/` | `*.sample` | Sample lists for REGENIE to use inconjuction with the `*.pheno` file |
| `summaries/` | `*.png` | Histograms and box-plots of phenotype disributions pre- and post-QC within strata |
| `summaries/` | `*.txt` | Summaries of phenotypes and ages pre and post-QC within strata. |
| `logs/` | `*_QC_individuals_removed.txt` | List of individuals removed because of failing phenotype QC |
| `logs/` | `*_removed_missing_data.txt` | List of individuals removed because missing value for phenotype or at least one covariate |
