---
title: "Phenotype Pipeline"
---
<div style="display: flex; justify-content: space-between; align-items: center;">
  <a href="./index.html">⬅️ Return to Homepage</a>
  <a href="./gwas.html">Go to Step 4 [Genome Wide Association Study] ➡️</a>
</div>

## Overview

This repository provides a containerized pipeline for QC and deriviation of phenotype data for subsequent GWAS analysis.

## Usage

1. **Edit `parameters.txt`** to set filenames and options for your data. This includes mapping column names to data labels that can be interpreted by the pipeline.

2. **Run the pipeline**

    ```
    # Run container using Apptainer
    bash PHENOTYPE_PIPELINE.sh --apptainer
    
    # Run container using Singulairty
    bash PHENOTYPE_PIPELINE.sh --singulairty
    ```

3. **Outputs** will be saved in the directory specified by `out_dir` in parameters.txt. Two sub-directories will be created:

  * `regenie_input/`
    * primary phenotype file
    * additional covariate file associated with genotyping / imputation process
    * subject lists to stratify individuals by ancestry and sex

  * `summaries/`
    * tabular summaries of phenotypes, split by each phenotype and ancestry
    * png file plotting histograms and box plots of distributions of pre-QC'd, QC'd, residuals, and rank normal values for GWAS


## Details

### 1. Files required
The script expects 3 files for input to be specified in the `parameters.txt` file.
```
## Filename containing phenotype and covariate data (required)
PHENO_COV_FILE="/path/to/PhenoCovars.txt"

## Filename containing ancestry labels [AFR|AMR|EAS|EUR|MID|SAS] (required)
ANCESTRY_FILE="/path/to/Ancestries.txt"

## Filename containing 20 study-level PCs generated by FLASH PCA (required)
PCS_FILE="/path/to/FlashPCAStudyPCA.txt"
```
##### Phenotype + Covariate File
This file should be a tab-delimited file with header that contains all variables for phenotype derivation for height, BMI and WHR, and all associated covariates. 

```
ID    Age   Sex  Height  Weight  Waist  Hip    CatVar1  CatVar2  CatVar3
P1    25    2    99      75.6    89     100    0        1        1
P2    85    1    170     89.8    91     98     1        1        2
P3    36    2    159     91.1    84     1020   0        2        3
...
```
##### Ancestry label file
This file should be a tab-delimited file with the header as shown with ancestry group assignment provided for each individual
```
IID     Ancestry
P1      EUR
P2      EUR
P3      EUR
...
```
##### Principal Components File
This file should be a tab-delimited file with 20 PCs (FLASHPCA output format)
```
IID   FID   PC1          PC2          PC3            ...    PC18         PC19         PC20          
P1    1     0.722770917  0.422132091  0.516251618    ...    0.05984563   0.74798664   0.310145635
P2    2     0.584095559  0.154674118  0.243537544    ...    0.059630973  0.708644846  0.863891645
P3    3     0.709639726  0.926419979  0.041459368    ...    0.613060395  0.315596588  0.719991385
...
```


### 2. Derivation of phenotypes
Where possible, please provide variables to derive phenotypes are required:

| Phenotype | Variables to priortise if available |
| :---: | :---: |
| **HEIGHT** | height (cm) |
| **BMI** | height (cm), weight (kg) |
| **WHR** | waist circumference (cm), hip circumference (cm) |

Note, WHR adjusted for BMI GWAS will be derived by the pipeline.

If you do not have measures of weight, waist and/or hip, pre-derived values for BMI and WHR can be used and these will be QC'd.

#### Quality control
Variables used to derive BMI and WHR will first undergo QC prior to derivation of BMI and WHR that will subsequently be QC'd again.

Details of quality control applied to variables can be found in the following table. Exclusions based on z-scores are performed within ancestry-sex.

| Variable / Phenotype | Range Inclusion | Z-score Inclusion |
| :---:       | :---:              | :---: |
| **HEIGHT**    | 100 - 220 (cm)    | -5 < z < 5 |
| **BMI**       | 15 - 100          | -5 < z < 5 |
| **WHR**       | None              | -5 < z < 5 |
| **WHRadjBMI** | None              | -5 < z < 5 |
| **WEIGHT**    | None              | -5 < z < 5 |
| **WAIST**     | 38.1 - 228.6 (cm) | -5 < z < 5 |
| **HIP**       | None              | None |
| **AGE**       | 18-110 (years)    | None |


#### Allowing for differences in timing when measurements taken
To account for the nature of EHR-based studies and other longitudinal studies, the pipeline allows for multiple measures of BMI and WHR to be provided. 

Specifically, two sets of variables may be provided for BMI if the variables used to derive primary BMI for GWAS and BMI used to adjust for WHR were measured at different times. 
Similary, two sets of variables may be provided for WHR if the variables used to derive primary WHR for GWAS and WHR adjusted for BMI  were measured at different times. 
As a result, the pipeline allows for different ages for height, BMI, WHR and WHRadjBMI to be provided.

